ref(ddm): streamline MRI naming and parsing

obostjancic opened the pull request:
- Implements https://www.notion.so/sentry/DDM-Frontend-Naming-e58cba9bbd444f67aa387d4dd70058bc
- Re-implements MRI parsing, Introduces MRI type
- Implements MRI field parsing and formatting

ArthurKnaus commented on static/app/types/metrics.tsx:
```diff
@@ -13,6 +11,19 @@ export type MetricsOperation =
   | 'p95'
   | 'p99';
 
+export type MetricType = 'c' | 'd' | 'g' | 'e' | 's';
+
+export type UseCase = 'custom' | 'transactions' | 'sessions';
+
+export type MRI = `${MetricType}:${UseCase}${string}@${string}`;
+
+export type ParsedMRI = {
+  name: string;
+  type: MetricType;
+  unit: string;
+  useCase: UseCase;
+};
+
```
love it ❤️

ArthurKnaus commented on static/app/utils/metrics/mri.tsx:
```diff
@@ -0,0 +1,93 @@
+export function parseField(field: string): {mri: MRI; op: string} | null {
+  const parsedFunction = parseFunction(field);
+  if (!parsedFunction) {
+    // We only allow aggregate functions for custom metric alerts
```
Probably not needed anymore in this context?

ArthurKnaus commented on static/app/utils/metrics/mri.tsx:
```diff
@@ -0,0 +1,93 @@
+export function getUseCaseFromMRI(mri?: string): UseCase {
+  const parsed = parseMRI(mri);
+
+  return parsed?.useCase ?? 'sessions';
```
Why is the default `sessions`? 
Shouldn't we just return `null` here and let the default value be the concern of the code that uses it?

ArthurKnaus commented on static/app/utils/metrics/mri.tsx:
```diff
@@ -0,0 +1,93 @@
+  return parsed?.mri ?? DEFAULT_MRI;
+}
+
+export function formatMRIAggregate(aggregate: string) {
```
```suggestion
export function formatMRIField(aggregate: string) {
```

ArthurKnaus commented on static/app/utils/metrics/mri.tsx:
```diff
@@ -0,0 +1,93 @@
+import {MetricType, MRI, ParsedMRI} from 'sentry/types/metrics';
+import {parseFunction} from 'sentry/utils/discover/fields';
+
+import {DEFAULT_METRIC_ALERT_AGGREGATE, UseCase} from './index';
```
should we move `DEFAULT_METRIC_ALERT_AGGREGATE` in here and rename it to `DEFAULT_METRIC_ALERT_FIELD`?

ArthurKnaus commented on static/app/views/alerts/rules/metric/utils/isCustomMetricAggregate.tsx:
```diff
@@ -1,6 +1,6 @@
-import {fieldToMri, getUseCaseFromMRI} from 'sentry/utils/metrics';
+import {getUseCaseFromMRI, parseField} from 'sentry/utils/metrics/mri';
 
 export const isCustomMetricAggregate = (aggregate: string): boolean => {
```
```suggestion
export const isCustomMetricField = (aggregate: string): boolean => {
```

ArthurKnaus commented on static/app/views/alerts/rules/metric/utils/isCustomMetricAlert.tsx:
```diff
@@ -1,8 +1,8 @@
  */
 export function isCustomMetricAlert(aggregate: string): boolean {
-  return fieldToMri(aggregate).mri !== undefined;
+  return Boolean(parseField(aggregate)?.mri);
```
nit: we prefer `!!`
```suggestion
  return !!parseField(aggregate)?.mri;
```

ArthurKnaus reviewed (approved):
Nice work! 🎉

matejminar commented on static/app/types/metrics.tsx:
```diff
@@ -59,12 +70,13 @@ export type MetricsTagValue = {
-  name: string;
+export type MetricMeta = {
+  mri: MRI;
+  // name is returned by the API but should not be used, use parseMRI(mri).name instead
+  // name: string;
```
Useful comment ❤️

matejminar commented on static/app/types/metrics.tsx:
```diff
@@ -13,6 +11,19 @@ export type MetricsOperation =
+
+export type UseCase = 'custom' | 'transactions' | 'sessions';
+
+export type MRI = `${MetricType}:${UseCase}${string}@${string}`;
```
🚀
I wonder what's the perf of this. Fingers crossed.

obostjancic commented:
@ArthurKnaus I think I addressed all of the comments in [8208e52](https://github.com/getsentry/sentry/pull/60543/commits/8208e525b9167c44e5e537e8643863ec40b1abb8)
